### PR TITLE
MODELIX-399 Lazy loading for the content explorer

### PR DESCRIPTION
**Pull request by tgorzsas**
Instead of loading the whole model when a repository is selected only the root node is loaded. 

image

Nodes on next levels are loaded when their parent node is expanded.

image

The Expand All button doesn't load the whole tree any more, but it expands one more level in the Content Explorer.

### CODE REVIEW

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt [hunk 1]
```diff
@@ -100,7 +131,22 @@ class ContentExplorer(private val client: IModelClient, private val repoManager:
         }
     }
 
-    private fun FlowContent.contentPageBody(rootNode: PNodeAdapter, versionHash: String) {
+    private fun collectParentNodesOneLevelDeeper(node: PNodeAdapter, expandedNodeIds: Set<String>): Set<String> {
+        if (expandedNodeIds.contains(node.nodeId.toString())) {
```
[the remark this answers is not in the document]
Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt:133]: This thread still needs an answer. Maybe some comment in the code would help.

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt [hunk 2]
```diff
@@ -100,7 +129,21 @@ class ContentExplorer(private val client: IModelClient, private val repoManager:
         }
     }
 
-    private fun FlowContent.contentPageBody(rootNode: PNodeAdapter, versionHash: String) {
+    private fun collectExpandableChildNodes(node: PNodeAdapter, expandedNodeIds: Set<String>): Set<String> {
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt:132]: The name is much better now. The idea of the arguments is that we look for nodes to expand under `node` and `expandedNodeIdes` is the state of nodes that are already expanded from previous interactions? If so, maybe the arguments can be renamed to better reflect this:

```suggestion
    private fun collectExpandableChildNodes(from: PNodeAdapter, alreadyExpandedNodeIds: Set<String>): Set<String> {
```

Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt:132]: or maybe `parent` or `under`?

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt [hunk 3]
```diff
@@ -70,17 +77,41 @@ class ContentExplorer(private val client: IModelClient, private val repoManager:
...
-                    bodyContent { contentPageBody(rootNode, versionHash) }
+                    bodyContent { contentPageBody(rootNode, versionHash, emptySet()) }
                 }
             }
+            post("/content/{versionHash}/") {
+                val versionHash = call.parameters["versionHash"]
+                if (versionHash.isNullOrEmpty()) {
+                    call.respondText("version not found", status = HttpStatusCode.InternalServerError)
+                    return@post
+                }
```
[the remark this answers is not in the document]
Comment (tgorzsas) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt]: Fixed, Bad Request is sent.

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt (hunk 2 above)
Comment (tgorzsas) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt]: Parameters are renamed.

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (tgorzsas) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt]: Some comments were added to the method to express its intent.

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt [hunk 4]
```diff
@@ -70,17 +77,41 @@ class ContentExplorer(private val client: IModelClient, private val repoManager:
...
+                    if (expandedNodes.expandAll)
+                        collectParentNodesOneLevelDeeper(rootNode, expandedNodes.expandedNodeIds)
+                    else
+                        expandedNodes.expandedNodeIds
+
+                call.respondText(
+                    buildString {
+                        appendHTML().ul("treeRoot") {
+                            nodeItem(rootNode, expandedNodeIds)
+                } })
```
[the remark this answers is not in the document]
Comment (tgorzsas) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt]: Formatting is fixed.

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/ContentExplorer.kt]: Thanks, that makes it a lot clearer now.

**Review by languitar — APPROVED** [review]

**Comment by languitar** [comment]
Let's wait for @mhuster23 to have a look on the changes before merging. From my side, this looks fine now.

**Review by mhuster23 — APPROVED** [review]

**Comment by mhuster23** [comment]
Seems like the commits aren't signed, so merging is blocked.

**Comment by tgorzsas** [comment]
> Seems like the commits aren't signed, so merging is blocked.

They are signed now.

**Pull request merged**